lib/components/room-link-validator.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * The room link validator is used to determine if a room can be bridged.
+ */
+const ConfigValidator = require("./config-validator");
+
+const VALIDATION_CACHE_LIFETIME = 30 * 60 * 1000;
+const PASSED = "RLV_PASSED";
+const ERROR = "RVL_ERROR";
+const ERROR_USER_CONFLICT = "RVL_USER_CONFLICT";
+const ERROR_CACHED = "RVL_ERROR_CACHED";
+
+const RULE_SCHEMA = {
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    type: "object",
+    properties: {
+        userIds: {
+            type: "object",
+            properties: {
+                exempt: {
+                    type: "array",
+                    items: {
+                        type: "string"
+                    }
+                },
+                conflict: {
+                    type: "array",
+                    items: {
+                        type: "string"
+                    }
+                }
+            }
+        }
+    }
+};
+
+class RoomLinkValidator {
+    /**
+     * @param {string} config Config for the validator.
+     * @param {string} config.ruleFile Filename for the rule file.
+     * @param {string} config.rules Rules if not using a rule file, will be
+     *                              overwritten if both is set.
+     * @param {boolean} config.watch If true, watch a rule file for changes.
+     * @param {AppServiceBot} asBot The AS bot.
+     */
+    constructor (config, asBot, log) {
+        this.log = log;
+        this.conflictCache = new Map(); // roomId => number
+        this.waitingRooms = new Map(); // roomId => Promise
+        this.asBot = asBot;
+
+        if (config.ruleFile) {
+            this.config = new ConfigValidator(RULE_SCHEMA);
+            this.readRuleFile(config.ruleFile);
+            if (config.watch) {
+                this.log(`Watching for rule changes on ${config.ruleFile}`);
+                this.config.watchForChanges(config.ruleFile, () => {
+                    try {
+                        this.readRuleFile(config.ruleFile);
+                    }
+                    catch (e) {
+                        // Catch these because we don't want to blow up in here.
+                    }
+                });
+            }
+        }
+        else if (config.rules) {
+            this.rules = this.evaulateRules(config.rules);
+        }
+        else {
+            throw new Error("Either config.ruleFile or config.rules must be set");
+        }
+    }
+
+    readRuleFile(filename) {
+        try {
```

Comment:
I think it's better to have the catch high up (where my comment about logging things is) and just have this throw.